### ISSUE TITLE
Tests failing because of different locales

### ISSUE THREAD

**Issue by cincuranet** [dotnet/efcore#14912]
Follow-up from #14889. The failing tests on Czech locale are these.

```text
DTO_complex_distinct_orderby(isAsync: False)
DTO_complex_distinct_orderby(isAsync: True)
DTO_complex_orderby(isAsync: False)
DTO_complex_orderby(isAsync: True)
OrderBy_Join(isAsync: True)
OrderBy_Join(isAsync: False)
OrderBy_Take_Last_gives_correct_result(isAsync: False)
OrderBy_Take_Last_gives_correct_result(isAsync: True)
OrderBy_Dto_projection_skip_take(isAsync: True)
OrderBy_Dto_projection_skip_take(isAsync: False)
DTO_member_distinct_orderby(isAsync: False)
DTO_member_distinct_orderby(isAsync: True)
```

Signalled by @lauxjpn in #20640: BuiltInDataTypesTestBase.Object_to_string_conversion

**Comment by smitpatel** [comment]
`Customer.City` column has certain characters with Unicode and difference in first letter's casing which gives ordering differences.

**Comment by ajcvickers** [comment]
@smitpatel Is the ordering happening because of client eval in the queries, or only on the L2O execution that we use to verify the results?

**Comment by smitpatel** [comment]
City has different orderings on server side too between different providers.

**Comment by roji** [comment]
@cincuranet can you confirm whether you're seeing culture-related failures on one of the built-in providers (SQL Server/Sqlclient) or only on Firebird? Can you post an exception or assertion failure, on, say, `OrderByJoin`?

**Comment by roji** [comment]
@cincuranet ping

**Comment by cincuranet** [comment]
Haven't had time to run the SQL Server tests yet. But it's on my list, don't worry.

**Comment by cincuranet** [comment]
@roji Yes, even the built-in providers are failing. I tried it quick'n'dirty in #15176. Which failed [here](https://dev.azure.com/dnceng/public/_build/results?buildId=135137), but when reverted went fine [here](https://dev.azure.com/dnceng/public/_build/results?buildId=135406).

**Comment by smitpatel** [comment]
The issue is L2O is using local culture but server never does. Perhaps we should specify invariant culture orderings for L2O query validation.

**Comment by ajcvickers** [comment]
Note from triage: bottom line here is that our tests should run on a non-English machine with a default installation of the database on that machine. It is expected that the database locale matches the machine locale. We should only force everything to English if really necessary.

**Comment by roji** [comment]
tl;dr we should consider forcing an invariant/ordinal collation on our Northwind test database, to avoid locale sensitivity.

The current state of our testing is that we force client-side ordinal ordering in tests, and expect the server to return matching results. This happens to work on an English locale SQL Server, but fails with others. We can try to switch to assuming server and client orderings are the same based on machine-configured locale, as suggested by  @ajcvickers above. This doesn't necessarily work as easily as we'd like:

* Sqlite does BINARY collation by default (which is why our tests pass), so we'd to to somehow make it work with the machine collation, which doesn't seem trivial/doable. The same would have to be done for PG, Oracle...
* If there's ever a provider whose collation behavior isn't exactly the same as .NET's for some culture value, this would break down.

So it seems overall safer to not depend on the machine locale, and to force our Northwind to have an invariant/ordinal database collation, which would make it match the .NET behavior. 

(I found some other minor locale-related issues, will submit fixes to those too).

**Comment by roji** [comment]
One more thought is that we can set the .NET locale to InvariantCulture globally, via a module initializer, rather than doing it in every place where we do a culture-sensitive operation (e.g. OrderBy). Each test assembly would need to do it separately.

**Comment by ErikEJ** [comment]
@roji would doing that not hide localization bugs in the product?

**Comment by roji** [comment]
@ErikEJ are you referring to the first proposal (set the database to invariant collation), or to the second (set .NET culture to invariant globally)? What exact bugs do you have in mind?

**Comment by ErikEJ** [comment]
Second. Back in the day I could find product bugs, by running the test suite on a machine with Danish regional settings - but maybe the analyzers take care of that now?

**Comment by roji** [comment]
OK. Yeah, at least simple cases such as StartsWith without culture should be caught by the analyzer - but maybe it's wiser to continue as we are, without setting globally.

**Comment by ErikEJ** [comment]
Or add a CI box with Danish culture??

**Comment by roji** [comment]
Design decision:

* There are two options here:
    1. Enforce some specific locale (e.g. English) both in the database and in the tests. This technically results in reduced coverage, since EF Core tests never get run with non-English locale. However, it means that for contributors, EF tests will never fail because of locale issues.
    2. Make sure that tests always pass if the client and database locales match. This technically helps coverage (contributors will be running with other locales, and so flush out possible locale-related bugs), but contributors where there's a client/database locale mismatch will get failing tests out of the box.
* We feel that we haven't discovered enough actual product bugs (as opposed to test bugs) to justify option 2, so we'll go with enforcing English everywhere. That means that test databases need to be created with an explicit locale, and all tests on the client must also be run in that locale.
* Once this is done, we can do a one-off run with another, non-English locale, to see if that uncovers any product bugs.

**Comment by ErikEJ** [comment]
@roji Back in the day I recall uncovering several test and some product bugs by running tests on my Danish system - but so be it.

Make sure to update the Build guide with instructions for developers using non-English locales.

**Comment by roji** [comment]
@ErikEJ

> Back in the day I recall uncovering several test and some product bugs by running tests on my Danish system - but so be it.

Yeah, we know we're giving up on some coverage - although the plan is to still do a run with a different non-English locale to see what that flushes out.

> Make sure to update the Build guide with instructions for developers using non-English locales.

One main point of going English-everywhere is precisely to avoid forcing developers to do anything, regardless of their locales. In other words, we'd cause English to be used both in the tests themselves (.NET locale) and in the test databases EF creates (e.g. SQL Server database collation).

**Comment by cincuranet** [comment]
Fixed by #33291.
